compiler/noirc_frontend/src/hir/type_check/mod.rs

```diff
@@ -36,7 +37,10 @@ pub struct TypeChecker<'interner> {
     /// verified at the end of a function. This is because constraints arise
     /// on each variable, but it is only until function calls when the types
     /// needed for the trait constraint may become known.
-    trait_constraints: Vec<(TraitConstraint, ExprId)>,
+    // trait_constraints: Vec<(TraitConstraint, ExprId)>,
+    // TODO, change this to a hashmap
+    // trait_constraints: HashMap<(TraitId, ExprId), TraitConstraint>,
+    trait_constraints: HashMap<ExprId, Vec<TraitConstraint>>,
```

Comment:
I also switched from a vec of all trait constraints to a hash map of ExprId -> Vec. I felt this was a clearer association between expressions and trait constraints, however, I can switch it back as I realized it is not necessary.